schema_context.go

```diff
@@ -0,0 +1,151 @@
+package main
+
+import (
+	"context"
+
+	jptr "github.com/qri-io/jsonpointer"
+)
+
+// SchemaContext holds the validation context
+// The aim is to have one global validation context
+// and use local sub contexts when evaluating parallel branches
+type SchemaContext struct {
+	Local                   *Schema
+	Root                    *Schema
+	RecursiveAnchor         *Schema
+	Instance                interface{}
+	LastEvaluatedIndex      int
+	LocalLastEvaluatedIndex int
+	BaseURI                 string
+	InstanceLocation        *jptr.Pointer
+	RelativeLocation        *jptr.Pointer
+	BaseRelativeLocation    *jptr.Pointer
+
+	LocalRegistry *SchemaRegistry
+
+	EvaluatedPropertyNames      map[string]bool
+	LocalEvaluatedPropertyNames map[string]bool
+	Misc                        map[string]interface{}
+
+	ApplicationContext *context.Context
+}
```

Comment:
Ok, after a bunch of reading, I think this API needs work, but seems to clearly point to how it can be improved.

* It seems to me this struct is a _state machine_, not a context. a "context" in go carries scope across API boundaries. This struct harmonizes state into a single location, allowing keywords to be stateless.
* because all fields are exported, the guts of this state machine are open to the world to modify
* the methods & fields on this struct are the primary API for keyword developers. Because we support custom keywords, It's a public API and needs to be written defensively.
* all keywords in this package are consumers of this API, and examples for other keyword developers, a massive win IMHO.
* this struct is not safe for concurrent use. That might be ok for now, but we should have a plan for making it safe for concurrency.
* the `Instance` couples Instance data to validation state, which feels incorrect. data should flow separately though the keyword API, especially in a streaming context
* this state machine should collect validation errors (we've been using an outParam for this)

Starting from the top level `Schema` "user api". Instead of being "a wrapper function to maintain some level of backwards compatibility with versions v0.1.2 and prior", let's break the API complete and define a very "generic" go function that initializes a root state and transitions to the "keyword API":

```go
func (s *Schema) Validate(ctx context.Context, data interface{}) {
	st := NewValidationState(s)
	s.ValidateKeyword(ctx, st, data)
}
```

The call to `ValidateKeyword` transitions us to the  "keyword API", where we've changed the primary  `ValidateFromContext` interface function to something like `ValidateKeyword`:

```go
// Keyword is an interface for anything that can validate.
// JSON-Schema keywords are all examples of Keyword
type Keyword interface {
  // ValidateKeyword runs a validation check against decoded JSON data, 
  // calling methods on ValidationState to record any discovered errors
  ValidateKeyword(ctx context.Context, state *ValidationState, data interface{})
  // ...
}
```

A keyword implementation would change to implement like this:
```go
// ValidateKeyword implements the Keyword interface for Maximum
func (m Maximum) ValidateKeyword(ctx context.Context, state *ValidateionState, data interface{}) {
	SchemaDebug("[Maximum] Validating")
	if num, ok := data.(float64); ok {
		if num > float64(m) {
			// state now keeps the errs slice internally, has all the info it needs to 
			// populate error fields
			state.AddError(fmt.Sprintf("must be less than or equal to %f", m))
		}
	}
}
```

a complex keyword needs a fairly rich API from the `state` struct. Here I've made up methods to satistify methods `Items` needs. Warning, untested code:

```go
// ValidateKeyword implements the Keyword interface for Items
func (it Items) ValidateKeyword(ctx context.Context, state *ValidationState, data interface{}) {
	SchemaDebug("[Items] Validating")
	if arr, ok := schCtx.Instance.([]interface{}); ok {
		// instead of "NewSchemaContextFromSourceClean(*schCtx)", state gets a method "subState"
		// that initializes & returns a clean substate from the parent
		subState := state.NewSubState()

		if it.single {
			// BaseRelativeLocation should be a method that reads from a private field,
			// the method defends against nil access, making it safe to use like this:
			if newPtr := state.BaseRelativeDescendant("items"); newPtr != nil {
				subState.SetBaseRelativeLocation(newPtr)
			}
			// this could probably be turned into a one-liner:
			newPtr := state.RelativeDescendant("items")
			subState.SetRelativeLocation(newPtr)
			for i, elem := range arr {
				if _, ok := state.LocalKeyword("additionalItems"); ok {
					state.SetPropertiesEvaluated("0")
					state.SetLocalPropertiesEvaluated("0")
					// These might be combined into some sort of "only increment if higher" setter
					if state.LastEvaluatedIndex() < i {
						state.SetEvaluatedIndex(i)
					}
					if state.LocalLastEvaluatedIndex() < i {
						state.SetLocalLastEvaluatedIndex(i)
					}
				}
				subState.ClearContext()
				newPtr = state.InstanceLocationDescendant(strconv.Itoa(i))
				subState.SetInstanceLocation(newPtr)
				// here it's clearer we're using a subState with a different data element:
				it.Schemas[0].ValidateKeyword(ctx, subState, elem)
				if _, ok := state.LocalKeyword("additionalItems"); ok {
					// TODO(arqu): this might clash with additionalProperties
					// should separate items out
					state.SetPropertiesEvaluated(subState.EvaluatedProperties()...)
					state.SetLocalPropertiesEvaluated(subState.LocalEvaluatedProperties()...)
				}
			}
		} else {
			for i, vs := range it.Schemas {
				if i < len(arr) {
					if _, ok := state.LocalKeyword("additionalItems"); ok {
						state.SetPropertyEvaluated(strconv.Itoa(i))
						state.SetLocalPropertyEvaluated(strconv.Itoa(i))
						// These might be combined into some sort of "only increment if higher" setter
						if state.LastEvaluatedIndex() < i {
							state.SetEvaluatedIndex(i)
						}
						if state.LocalLastEvaluatedIndex() < i {
							state.SetLocalLastEvaluatedIndex(i)
						}
					}
					subState.ClearContext()
					if newPtr := state.BaseRelativeDescendant("items", strconv.Itoa(i)); newPtr != nil {
						subState.SetBaseRelativeLocation(newPtr)
					}
					newPtr, _ := state.RelativeLocationDescendant("items", strconv.Itoa(i))
					subState.SetRelativeLocation(newPtr)
					newPtr = state.InstanceLocationDescendant(strconv.Itoa(i))
					subState.SetInstanceLocation(newPtr)

					vs.ValidateKeyword(ctx, subState, arr[i])
					if _, ok := state.LocalKeyword("additionalItems"); ok {
						state.SetPropertiesEvaluated(subState.EvaluatedProperties()...)
						state.SetLocalPropertiesEvaluated(subState.LocalEvaluatedProperties()...)
					}
				}
			}
		}
	}
}
```

The hard work of figuring out how to arrange this is done, all of these comments are just ergonomics, but important considering we're going to break the API. I think breaking the API is wholly appropriate with the transition to the `2019_09` spec.